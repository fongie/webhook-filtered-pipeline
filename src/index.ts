import { BuildSpec, FilterGroup, ISource, Project, Source } from "@aws-cdk/aws-codebuild";
import { Pipeline, PipelineProps } from "@aws-cdk/aws-codepipeline";
import { GitHubSourceAction, GitHubSourceActionProps, GitHubTrigger } from "@aws-cdk/aws-codepipeline-actions";
import { Effect, PolicyStatement } from "@aws-cdk/aws-iam";
import { Construct } from "constructs";

/**
 * Properties for the webhook filtered pipeline.
 * You should either use 'githubSourceActionProps' together with 'webhookFilters'
 * to create a pipeline with a GitHub source. Then, the pipeline will have the source action
 * set. Or, you provide your own 'source' for the codebuild project. Then, you will also have
 * to set the pipeline source action manually after creating the pipeline.
 */
export interface WebhookFilteredPipelineProps extends PipelineProps {
  /**
   * GitHub details used to set the starting project source as well as the first action
   * in the pipeline. Remember to set your output artifacts etc here to keep building
   * the pipeline.
   * You must set this and webhookfilters OR source.
   */
  readonly githubSourceActionProps?: GitHubSourceActionProps;
  /**
   * All webhook filters you wish to apply to starting the pipeline
   * You must set this and githubSourceActionProps OR source.
   */
  readonly webhookFilters?: FilterGroup[];

  /**
   * If you do not wish to use GitHub, you can provide a custom
   * source here. Note that if you do, you have to set your own
   * filters using `source.webhookFilters` and you must set
   * `source.webhook` to true. Also, you have to set the source action
   * on the pipeline after creating it.
   * 
   * NOTE: if you set this, all github properties (including webhookFilters) 
   * above will be ignored.
   */
  readonly source?: ISource;
}

/**
 * Start an AWS CodePipeline using a source with webhook filters
 * This is currently not supported in CloudFormation so this construct
 * uses a workaround using a CodeBuild project. 
 * See issue https://github.com/aws/aws-cdk/issues/10265
 * 
 * If you set the github props, you will get a pipeline with the first source
 * stage configured. If you set the 'source' prop, you will need to confgure the
 * pipeline's source stage yourself.
 */
export class WebhookFilteredPipeline extends Pipeline {
  private _starterProject: Project
  constructor(scope: Construct, id: string, props: WebhookFilteredPipelineProps) {
    super(scope, id, props);
    const { source, webhookFilters, githubSourceActionProps } = props;
    if (!source && (!githubSourceActionProps || !webhookFilters)) {
      throw Error('You must set EITHER source OR githubSourceActionProps and webhookfilters');
    }
    this._starterProject = new Project(this, `${id}-starter`, {
      source: source ?? Source.gitHub({
        owner: githubSourceActionProps.owner,
        repo: githubSourceActionProps.repo,
        branchOrRef: githubSourceActionProps.branch,
        webhook: webhookFilters ? true : false,
        webhookFilters
      }),
      buildSpec: BuildSpec.fromObject({
        version: '0.2',
        phases: {
          install: {
            'runtime-versions': {
              nodejs: 10
            }
          },
          build: {
            commands: [`aws codepipeline start-pipeline-execution --name ${this.pipelineName}`]
          }
        }
      })
    })
    this._starterProject.addToRolePolicy(new PolicyStatement({
      actions: [ 'codepipeline:StartPipelineExecution' ],
      resources: [ this.pipelineArn ],
      effect: Effect.ALLOW
    }));
    if (githubSourceActionProps) {
      this.addStage({
        stageName: 'Source',
        actions: [new GitHubSourceAction({
          ... githubSourceActionProps,
          trigger: GitHubTrigger.NONE // so it is only started by the starter project
        })],
      });
    }
  }
  /**
   * The CodeBuild project that starts the pipeline
   */
  get starterProject(): Project {
    return this._starterProject
  }
}
