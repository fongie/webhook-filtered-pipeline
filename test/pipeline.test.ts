import { BuildSpec, EventAction, FilterGroup, Project } from '@aws-cdk/aws-codebuild';
import { Artifact } from '@aws-cdk/aws-codepipeline';
import { CodeBuildAction } from '@aws-cdk/aws-codepipeline-actions';
import { App, SecretValue, Stack } from '@aws-cdk/core';
import { WebhookFilteredPipeline } from '../src';

let app: App;
let stack: Stack;
beforeEach(() => {
  app = new App();
  stack = new Stack(app);
});

test('throws on invalid props', () => {
  expect(() => new WebhookFilteredPipeline(stack, 'test', {})).toThrowError();
  expect(() => new WebhookFilteredPipeline(stack, 'test', {
    webhookFilters: [],
  })).toThrowError();
  expect(() => new WebhookFilteredPipeline(stack, 'test', {
    githubSourceActionProps: {
      owner: 'test',
      repo: 'testrepo',
      output: new Artifact(),
      oauthToken: SecretValue.plainText('mytoken'),
      actionName: 'github',
    },
  })).toThrowError();
});

test('snapshot for github', () => {
  const sourceOutput = new Artifact();
  const pipe = new WebhookFilteredPipeline(stack, 'test', {
    githubSourceActionProps: {
      owner: 'test',
      repo: 'testrepo',
      output: sourceOutput,
      oauthToken: SecretValue.plainText('mytoken'),
      actionName: 'github',
    },
    webhookFilters: [
      FilterGroup.inEventOf(EventAction.PUSH).andBranchIs('mybranch').andFilePathIsNot('*.md'),
    ],
  });
  pipe.addStage({
    stageName: 'stage2',
    actions: [new CodeBuildAction({
      input: sourceOutput,
      actionName: 'build',
      project: new Project(stack, 'proj', { buildSpec: BuildSpec.fromObject({}) }),
    })],
  });
  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});