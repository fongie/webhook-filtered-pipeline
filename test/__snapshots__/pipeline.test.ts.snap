// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for github 1`] = `
Object {
  "Resources": Object {
    "proj801E7F7D": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "NO_ARTIFACTS",
        },
        "Cache": Object {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": Object {
          "Fn::GetAtt": Array [
            "testArtifactsBucketEncryptionKey8916362F",
            "Arn",
          ],
        },
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:1.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "projRole5BB03F29",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "{}",
          "Type": "NO_SOURCE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "projRole5BB03F29": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "projRoleDefaultPolicy9A692AF1": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "proj801E7F7D",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "proj801E7F7D",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    Object {
                      "Ref": "proj801E7F7D",
                    },
                    "-*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testArtifactsBucket6F90E1EE",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testArtifactsBucket6F90E1EE",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:DescribeKey",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testArtifactsBucketEncryptionKey8916362F",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testArtifactsBucketEncryptionKey8916362F",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "projRoleDefaultPolicy9A692AF1",
        "Roles": Array [
          Object {
            "Ref": "projRole5BB03F29",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testAF53AC38": Object {
      "DependsOn": Array [
        "testRoleDefaultPolicy0681E92B",
        "testRoleA48AD45D",
      ],
      "Properties": Object {
        "ArtifactStore": Object {
          "EncryptionKey": Object {
            "Id": Object {
              "Fn::GetAtt": Array [
                "testArtifactsBucketEncryptionKey8916362F",
                "Arn",
              ],
            },
            "Type": "KMS",
          },
          "Location": Object {
            "Ref": "testArtifactsBucket6F90E1EE",
          },
          "Type": "S3",
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "testRoleA48AD45D",
            "Arn",
          ],
        },
        "Stages": Array [
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Source",
                  "Owner": "ThirdParty",
                  "Provider": "GitHub",
                  "Version": "1",
                },
                "Configuration": Object {
                  "Branch": "master",
                  "OAuthToken": "mytoken",
                  "Owner": "test",
                  "PollForSourceChanges": false,
                  "Repo": "testrepo",
                },
                "Name": "github",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "Artifact_Source_github",
                  },
                ],
                "RunOrder": 1,
              },
            ],
            "Name": "Source",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ProjectName": Object {
                    "Ref": "proj801E7F7D",
                  },
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "Artifact_Source_github",
                  },
                ],
                "Name": "build",
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "teststage2buildCodePipelineActionRole62358A1E",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "stage2",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "testArtifactsBucket6F90E1EE": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "KMSMasterKeyID": Object {
                  "Fn::GetAtt": Array [
                    "testArtifactsBucketEncryptionKey8916362F",
                    "Arn",
                  ],
                },
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testArtifactsBucketEncryptionKey8916362F": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
                "kms:GenerateDataKey",
                "kms:TagResource",
                "kms:UntagResource",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "testRoleA48AD45D",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:DescribeKey",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "projRole5BB03F29",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "projRole5BB03F29",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Delete",
    },
    "testArtifactsBucketEncryptionKeyAlias6E1C67C0": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AliasName": "alias/codepipeline-test",
        "TargetKeyId": Object {
          "Fn::GetAtt": Array [
            "testArtifactsBucketEncryptionKey8916362F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
      "UpdateReplacePolicy": "Delete",
    },
    "testArtifactsBucketPolicyA3D14421": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "testArtifactsBucket6F90E1EE",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testArtifactsBucket6F90E1EE",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testArtifactsBucket6F90E1EE",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "testRoleA48AD45D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testRoleDefaultPolicy0681E92B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testArtifactsBucket6F90E1EE",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testArtifactsBucket6F90E1EE",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testArtifactsBucketEncryptionKey8916362F",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "teststage2buildCodePipelineActionRole62358A1E",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testRoleDefaultPolicy0681E92B",
        "Roles": Array [
          Object {
            "Ref": "testRoleA48AD45D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "teststage2buildCodePipelineActionRole62358A1E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "teststage2buildCodePipelineActionRoleDefaultPolicyD6A49327": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "proj801E7F7D",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "teststage2buildCodePipelineActionRoleDefaultPolicyD6A49327",
        "Roles": Array [
          Object {
            "Ref": "teststage2buildCodePipelineActionRole62358A1E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testteststarterCB2336CC": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "NO_ARTIFACTS",
        },
        "Cache": Object {
          "Type": "NO_CACHE",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:1.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "testteststarterRole3087A756",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": Object {
            "Fn::Join": Array [
              "",
              Array [
                "{
  \\"version\\": \\"0.2\\",
  \\"phases\\": {
    \\"install\\": {
      \\"runtime-versions\\": {
        \\"nodejs\\": 10
      }
    },
    \\"build\\": {
      \\"commands\\": [
        \\"aws codepipeline start-pipeline-execution --name ",
                Object {
                  "Ref": "testAF53AC38",
                },
                "\\"
      ]
    }
  }
}",
              ],
            ],
          },
          "Location": "https://github.com/test/testrepo.git",
          "ReportBuildStatus": true,
          "Type": "GITHUB",
        },
        "Triggers": Object {
          "FilterGroups": Array [
            Array [
              Object {
                "Pattern": "PUSH",
                "Type": "EVENT",
              },
              Object {
                "Pattern": "refs/heads/mybranch",
                "Type": "HEAD_REF",
              },
              Object {
                "ExcludeMatchedPattern": true,
                "Pattern": "*.md",
                "Type": "FILE_PATH",
              },
            ],
          ],
          "Webhook": true,
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "testteststarterRole3087A756": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testteststarterRoleDefaultPolicy96648F66": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "testteststarterCB2336CC",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "testteststarterCB2336CC",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    Object {
                      "Ref": "testteststarterCB2336CC",
                    },
                    "-*",
                  ],
                ],
              },
            },
            Object {
              "Action": "codepipeline:StartPipelineExecution",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codepipeline:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    Object {
                      "Ref": "testAF53AC38",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testteststarterRoleDefaultPolicy96648F66",
        "Roles": Array [
          Object {
            "Ref": "testteststarterRole3087A756",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
